### ISSUE TITLE
Data export & 3rd party integrations

### ISSUE THREAD

**Issue by HelmiRifai** [metabase/metabase#2309]
I've been using Metabase for a few weeks while setting up a new analytics and visualization platform at work.

From the main Github page I see the following https://github.com/metabase/metabase/#extending-and-deep-integrations

I couldn't find how to go ahead doing that exactly yet. Ideally I wan't to be able to export query results ( & modify them for export as well [de-aggregate them for example]). This would be used by individuals in the Marketing and BI departments.

As an example let's say the marketing team is interested in launching an Ad campaign on Facebook for users that did a certain action in the past month. Right now I can show them the aggregated and divide them by time and so on, but I would like to be able to extend the current implementation to allow the team to take the results (inside the aggregation) and export them, either to files or directly call 3rd party APIs like Facebook Marketing API. They would get a button next to the Download CSV button that would take them through the process of exporting the data.

Any tips on going through with this would be appreciated.

**Comment by salsakran** [comment]
You can download the results of any query by clicking on the download button  - 
<img width="37" alt="screen shot 2016-04-05 at 2 47 34 pm" src="https://cloud.githubusercontent.com/assets/107439/14299060/37d48bde-fb3d-11e5-9366-a012a9c3e1b6.png">

Regarding more involved integration, you'd need to log in and run queries through the query api.

**Comment by HelmiRifai** [comment]
I mentioned the current download as CSV button in my original post. Your second point is to log in and run queries through the query api, do you mean log in on the current front end interface or just to call /api/ methods available? I feel you missed the gist of my post.

**Comment by tlrobinson** [comment]
@HelmiRifai currently in order to add new UI elements to Metabase, such as a button to export to an external service, you'll need to modify the React frontend and recompile Metabase. If you're interested in that I can give you some guidance.

If you'd like to use Metabase's query engine from an external application you can hit the HTTP API directly. http://discourse.metabase.com/t/external-http-api/175 We don't have great API docs at the moment but you can take a look at the API endpoints here https://github.com/metabase/metabase/tree/master/src/metabase/api and watch the requests in you browser's network inspector to get an idea of how it works.

Either way I'm curious about the types of customizations users are looking for. If you think this is something we could make easier let us know what you envision.

**Comment by HelmiRifai** [comment]
@tlrobinson Hey! Thanks for the detailed post. Yes that is what I am trying to do, I reached the point where I found the react components and had a look at them; I am not versed in React yet but I will definitely try to tweak there and recompile.

I also did find the API end points from https://github.com/metabase/metabase/tree/master/src/metabase/api and was thinking of doing something like /api/dataset/csv with /api/dataset/export/{id} where id would then handle the 3rd Party API for that id and do something like export the user advertising id's to Facebook Marketing API where an audience will be created and then our Marketing team can go ahead and handle the campaigns. Facebook will be my starting project in terms of integrations and customization.

Will look into the discourse link for the api.

Need to set up the dev environment and dig into the codebase now, just have to manage to get the time between other tasks I need to do, I am willing to collaborate on this if possible.

Thanks for the detailed response again!
